apps/web/src/components/Publication/OpenAction/UnknownModule/Decent/CurrencySelector.tsx

```diff
@@ -0,0 +1,198 @@
+import type { FC } from 'react';
+import type { Address } from 'viem';
+
+import { STATIC_IMAGES_URL } from '@hey/data/constants';
+import getRedstonePrice from '@hey/helpers/getRedstonePrice';
+import getTokenImage from '@hey/helpers/getTokenImage';
+import { useQuery } from '@tanstack/react-query';
+import { useAllowedTokensStore } from 'src/store/persisted/useAllowedTokensStore';
+import { formatUnits } from 'viem';
+import { useAccount, useBalance } from 'wagmi';
+
+const OA_SUPPORTED_CURRENCIES = ['USDC', 'WETH', 'WMATIC'];
+
+interface CurrencySelectorProps {
+  onSelectCurrency: (currency: Address) => void;
+}
+
+const CurrencySelector: FC<CurrencySelectorProps> = ({ onSelectCurrency }) => {
+  const { allowedTokens } = useAllowedTokensStore();
+  const { address } = useAccount();
+
+  const { data: wmaticBalanceData, isLoading: wmaticBalanceLoading } =
+    useBalance({
+      address,
+      chainId: 137,
+      query: { refetchInterval: 10000 },
+      token: allowedTokens.find((token) => token.symbol === 'WMATIC')
+        ?.contractAddress as Address
+    });
+
+  const { data: wethBalanceData, isLoading: wethBalanceLoading } = useBalance({
+    address,
+    chainId: 137,
+    query: { refetchInterval: 10000 },
+    token: allowedTokens.find((token) => token.symbol === 'WETH')
+      ?.contractAddress as Address
+  });
+
+  const { data: usdcBalanceData, isLoading: usdcBalanceLoading } = useBalance({
+    address,
+    chainId: 137,
+    query: { refetchInterval: 10000 },
+    token: allowedTokens.find((token) => token.symbol === 'USDC')
+      ?.contractAddress as Address
+  });
+
+  const { data: wmaticPriceUsd, isLoading: wmaticPriceLoading } = useQuery({
+    enabled: Boolean(wmaticBalanceData),
+    queryFn: async () => await getRedstonePrice('MATIC'),
+    queryKey: ['getRedstonePrice', 'WMATIC']
+  });
+
+  const { data: wethPriceUsd, isLoading: wethPriceLoading } = useQuery({
+    enabled: Boolean(wethBalanceData),
+    queryFn: async () => await getRedstonePrice('ETH'),
+    queryKey: ['getRedstonePrice', 'WETH']
+  });
+
+  const { data: usdcPriceUsd, isLoading: usdcPriceLoading } = useQuery({
+    enabled: Boolean(usdcBalanceData),
+    queryFn: async () => await getRedstonePrice('USDC'),
+    queryKey: ['getRedstonePrice', 'USDC']
+  });
+
+  const balances = {
+    USDC:
```

Comment:
better we can have switch case? i see lot of dupes here

Reply:
I think it can be removed if we use fiat from allowedTokens? no?

Reply:
refer to the changes done by @defispartan

> Re-wrote this component to use useRatesStore for USD prices, add new helper function to format token and usd amounts, and query tokens based on addresses instead of symbol